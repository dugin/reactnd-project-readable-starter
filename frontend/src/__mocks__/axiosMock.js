import axios from 'axios';
import MockAdapter from 'axios-mock-adapter';
import {categories, posts} from "./responseMock";

const mock = new MockAdapter(axios);

const URL = 'http://localhost:3001';

mock.onGet(URL + '/categories').reply(200, categories);

mock.onGet(URL + '/posts').reply(200, posts);
