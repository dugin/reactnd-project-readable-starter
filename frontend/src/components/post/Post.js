import React, {Component} from 'react';
import './Post.css';
import IosComposeOutline from 'react-icons/lib/io/ios-compose-outline';
import IosPersonOutline from 'react-icons/lib/io/ios-person-outline';
import Trophy from 'react-icons/lib/io/trophy';
import moment from 'moment';
import {Link} from 'react-router-dom';

class Post extends Component {

    componentWillReceiveProps(nextProps) {
    }

    setDate = () => {
        return moment(this.props.post.timestamp).format('MMM, DD - h:mm a')
    };

    render() {
        return (
            <div className="Post">
                <div className="Post__main">
                    <div className="Post__main__score">
                        <Trophy size={25}/>
                        {this.props.post.voteScore}
                    </div>
                    <div className="Post__main__info">
                        <Link to={'post/'+this.props.post.id} className="Post__main__info__title">  {this.props.post.title}</Link>
                        <h3 className="Post__main__info__body"> {this.props.post.body}</h3>
                    </div>
                </div>
                <div className="Post__more">
                    <div><IosComposeOutline size={25}/>{this.setDate()}</div>
                    <div><IosPersonOutline size={25}/>{this.props.post.author}</div>
                    <div className="badge badge-pill badge-secondary">{this.props.post.category}</div>

                </div>
            </div>
        )
    }
}

export default Post;