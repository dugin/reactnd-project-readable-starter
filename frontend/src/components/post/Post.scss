@import '../../variables';

.Post {

  margin-top: -1px;
  margin-bottom: 5px;
  background-color: white;
  padding: 10px;
  border: 1px solid $line-color;

  &__main {
    display: flex;

    &__score {
      display: flex;
      flex-direction: column;
      align-items: center;
      font-size: 14px;
      border: 1px solid map-get($theme-colors, success);
      color: $green;
      border-radius: 50%;
      padding: 12px 25px;
    }

    &__info {
      margin-left: 10px;

      &__title {
        font-size: 14px;
        color: $gray-900;

        &:hover{
          color: map-get($theme-colors, success);
      }

        @media (min-width: map-get($grid-breakpoints, sm)) {
          font-size: 20px;
        }
      }

      &__body {
        font-size: 12px;
        text-align: justify;
        color: $gray-500;
        overflow: hidden;
        text-overflow: ellipsis;
        display: -webkit-box;
        /*! autoprefixer: off */
        -webkit-box-orient: vertical;
        /*! autoprefixer: on */
        -webkit-line-clamp: 3;

        @media (min-width: map-get($grid-breakpoints, sm)) {
          font-size: 14px;
        }
      }
    }
  }

  &__more {
    margin-top: 10px;
    display: flex;
    justify-content: space-between;
    align-items: center;
    font-size: 12px;
    color: $gray-700;

    @media (min-width: map-get($grid-breakpoints, sm)) {
      font-size: 14px;
    }

    .badge {
      line-height: unset;
      font-size: 12px;

      @media (min-width: map-get($grid-breakpoints, sm)) {
        font-size: 14px;
      }
    }
  }
}