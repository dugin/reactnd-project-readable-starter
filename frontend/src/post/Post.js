import React, {PureComponent} from 'react';
import AddSubtract from "../components/AddSubtract";
import styled from 'styled-components';
import colors from "./../styles/constants/colors";
import {StyledPrimaryChip} from './../styles/styles';
import PropTypes from "prop-types";
import moment from 'moment';
import {voteType} from "../utils/constants";
import {connect} from "react-redux";
import {removePost, voteOnPost} from "./post.action";
import OptionMenu from "../components/OptionMenu";
import {withRouter} from "react-router-dom";
import RemoveDialog from "../components/RemoveDialog";

const StyledSection = styled.section`
 background-color: white;
  padding: 15px;
`

const StyledDiv = styled.section`
    display: flex;
    align-items: center;
   
    p{
    font-size: 14px;
    }
`;
const StyledSecondaryChip = StyledPrimaryChip.extend`
   background-color: ${colors.secondary};
`;


const StyledOptions = styled.section`
    display: flex;
    flex-direction: column;
    align-items: center;
`;

const StyledMain = styled.section`
   flex-grow: 1;
`;
const StyledFooter = styled.section`
   display: flex;
   align-items: center;
   margin-left: -10px;
   p{
      margin: 0 0 0 10px;
   }
`;

const Title = ({title}) => {
    return (
        <h1>
            {title}
        </h1>
    )
};

const Body = ({body}) => {
    return (
        <p>
            {body}
        </p>
    )
};


const Footer = ({author, timestamp, voteScore}) => {

    const fromNow = () => {
        return moment(timestamp).fromNow();
    };

    return (
        <StyledFooter>
            <StyledSecondaryChip
                label={voteScore}
            />
            <p>by <b>{author}</b></p>
            <p>{fromNow()}</p>

        </StyledFooter>
    )
};

class Post extends PureComponent {

    state = {
        openRemoveDialog: false
    };

    onAddOrSubtract = (voteType) => {
        this.props.dispatch(voteOnPost(this.props.post.id, voteType))
    };

    onOption = (o) => {

        if (o === 'edit') {
            this.props.history.push(`/edit/${this.props.post.id}`);
        }

        else if (o === 'remove') {
            this.setState({openRemoveDialog: true});
        }

    };

    onRemove = () => {
        this.props.dispatch(removePost(this.props.post.id));
    };

    render() {
        const {post} = this.props;

        return (
            <StyledSection>
                <StyledDiv>
                    <div>
                        <AddSubtract
                            onAddOrSubtract={this.onAddOrSubtract}
                        />
                    </div>

                    <StyledMain>
                        <Title
                            title={post.title}
                        />
                        <Body
                            body={post.body}
                        />
                    </StyledMain>

                    <StyledOptions>
                        <OptionMenu
                            options={['edit', 'remove']}
                            onSelect={(o) => this.onOption(o)}
                        />
                        <StyledPrimaryChip
                            label={post.category}
                        />
                    </StyledOptions>
                </StyledDiv>

                <Footer
                    author={post.author}
                    timestamp={post.timestamp}
                    voteScore={post.voteScore || 1}
                />

                <RemoveDialog
                    onClose={() => this.setState({openRemoveDialog: false})}
                    onRemove={this.onRemove}
                    open={this.state.openRemoveDialog}
                    post={post}
                />
            </StyledSection>
        );
    }
}

Post.propTypes = {
    post: PropTypes.object.isRequired
};
export default withRouter(connect()(Post));