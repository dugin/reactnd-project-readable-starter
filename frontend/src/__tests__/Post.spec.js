import React from 'react';
import toJson from 'enzyme-to-json';
import {shallow} from 'enzyme';
import Post from "../components/post/Post";
import {posts} from "../__mocks__/responseMock";

describe('Post Component', () => {

    const post = shallow(<Post post={posts[0]}/>);

    it('renders properly', () => {

        expect(toJson(post)).toMatchSnapshot();

    });
});