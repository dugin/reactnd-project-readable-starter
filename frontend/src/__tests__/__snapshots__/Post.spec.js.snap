// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Post Component renders properly 1`] = `
<div
  className="Post"
>
  <div
    className="Post__main"
  >
    <div
      className="Post__main__score"
    >
      <IoTrophy
        size={25}
      />
      6
    </div>
    <div
      className="Post__main__info"
    >
      <Link
        className="Post__main__info__title"
        replace={false}
        to="post/8xf0y6ziyjabvozdd253nd"
      >
          
        Udacity is the best place to learn React
      </Link>
      <h3
        className="Post__main__info__body"
      >
         
        Everyone says so after all.
      </h3>
    </div>
  </div>
  <div
    className="Post__more"
  >
    <div>
      <IoIosComposeOutline
        size={25}
      />
      Jun, 28 - 11:21 pm
    </div>
    <div>
      <IoIosPersonOutline
        size={25}
      />
      thingtwo
    </div>
    <div
      className="badge badge-pill badge-secondary"
    >
      react
    </div>
  </div>
</div>
`;
